design/pipeline-triggers.md

```diff
@@ -0,0 +1,369 @@
+
+At high level there are 3 different types of pipeline triggers.
+1. Resource triggers
+2. Schedule triggers
+3. Webhook triggers
+
+## Resource triggers
+You can enable triggers on the resources defined in your pipeline. Resources can be of types pipelines, repositories, containers and packages.
+Triggers are enabled by default on all the resources. However, you can choose to override/disable triggers for each resource.
+
+
+
+### Pipelines
+A new pipeline is triggered automatically whenever a new run of the `pipeline` resource is succesfully completed. See [pipeline resources](pipeline-resources.md#resources-pipelines) for more details.
+
+#### Scenarios
+- I would like to trigger my pipeline when a new ‘helm-package’ artifact is published by ‘Helm-CI’ pipeline that ran on `releases/*` branch.
+- I would like to trigger my pipeline when a new ‘helm-package’ artifact is published and tested as part of Helm-CI pipeline and tagged as 'Production'.
+- I would like to trigger my pipeline when ‘TFS-Update’ pipeline has completed ‘Ring2’ stage.  
+
+#### Schema
+```yaml
+resources:       
+  pipelines:
+  - pipeline: string 
+    source: string  
+    trigger:     # Optional; Triggers are enabled by default.
+      branches:  # branch conditions to filter the events, optional; Defaults to all branches.
+        include: [ string ]  # branches to consider the trigger events, optional; Defaults to all branches.
+        exclude: [ string ]  # branches to discard the trigger events, optional; Defaults to none.
+      stages: [ string ]  # trigger after completion of given stage, optional; Defaults to all stage completion.
+      tags: [ string ]  # tags on which the trigger events are considered, optional; Defaults to any tag or no tag.
+```
+
+#### Examples
+Triggers are enabled by default unless expliciltly opted out. You can disable the triggers on the `pipeline` resource.
+```yaml
+resources:
+  pipelines:
+  - pipeline: SmartHotel
+    source: SmartHotel-CI 
+    trigger: none   
+```
+
+
+You can control which branches get the triggers with a simple syntax.
+```yaml
+resources:
+  pipelines:
+  - pipeline: SmartHotel
+    source: SmartHotel-CI 
+    trigger: 
+    - releases/*
+    - master
+ ```
+ You can specify the full name of the branch (for example, master) or a prefix-matching wildcard (for example, releases/*). You cannot put a wildcard in the middle of a value. For example, releases/*2018 is invalid.
+ 
+
+
+You can specify the branches to include and exclude.
+```yaml
+resources:
+  pipelines:
+  - pipeline: SmartHotel
+    source: SmartHotel-CI 
+    trigger: 
+      branches:
+        include: 
+        - releases/*
+        exclude:
+        - master
+ ```
+ 
+ 
+ You can specify which tags to control the triggers.
+ ```yaml
+resources:
+  pipelines:
+  - pipeline: SmartHotel
+    source: SmartHotel-CI 
+    trigger: 
+      branches:
+        include: 
+        - releases/*
+      tags: 
```

Comment:
are the tags and'd or or'd ? Can you clarify in the documentation?

Reply:
I may not fully understand the scenarios, so feel free to educate me. The behavior I'd naively expect is to OR everything. The "only if production && signed" case seems like an advanced scenario. Two ideas from me:

- Suggest a pattern of "put a step up front that checks for your custom conditions, halt the job if it doesn't match".
- Offer an alternate syntax that goes something like:
```yaml
...
  trigger:
    custom: >
      and(
        branchMatch(resources.foo.branch, 'releases/*'),
        eq(resources.foo.tag, 'Signed')
      )
```
(Obviously I just invented the `branchMatch` function; we would need to design the right set of functions here.)

Reply:
Ha, silly me. Thanks Eric.

Reply:
Discussed with @vtbassmatt. 
The inputs for tags/branches/stages are OR'd. 
For `AND` cases, we will introduce custom condition support inside trigger which takes expression input. This seems to be better and more powerful approach. This notation is already known to customers as part of step conditions.Later, when we introduce webhook triggers, this custom condition will be useful.

We need to define the functions appropriately. I will get back with the schema for this. However, since 'AND' is a corner case, we might pick this work in the next iteration.